### PR TITLE
Feat/worker

### PR DESCRIPTION
**Pull request by jackey8616**

### CODE REVIEW

File: src/runner/runner.module.ts [hunk 1]
```diff
@@ -1,23 +1,65 @@
...
+        ConfigModule.forRoot({
+          envFilePath: `${cwd()}/.env${nodeEnv ? `.${nodeEnv}` : ''}`,
+          isGlobal: true,
+        }),
+        BullModule.forRootAsync({
+          imports: [ConfigModule],
+          useFactory: (configService: ConfigService<{
+            QUEUE_REDIS_URI: string;
+          }>) => {
+            const queueRedisUri = configService.getOrThrow('QUEUE_REDIS_URI');
```
Comment (kkshyu) [inline src/runner/runner.module.ts:39]: how about using constant string to prevent from typo?

File: src/utility/lock/distribute_lock.service.ts [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+  constructor(
+    private cacheService: CacheService,
+  ) {}
+
+  async acquireLock(
+    key: string, acquirerUuid: string, expireTime?: number,
+  ): Promise<boolean> {
+    try {
+      const redisCli = this.cacheService.getClient();
+      const previousAcquired = await redisCli.get(`lock:${key}`);
```
Comment (kkshyu) [inline src/utility/lock/distribute_lock.service.ts]: it should be `previousAcquirerUuid`

File: src/utility/lock/distribute_lock.service.ts [hunk 3]
```diff
@@ -0,0 +1,66 @@
+import { Injectable } from '@nestjs/common';
+
+import { CacheService } from '../cache/cache.service';
+
+@Injectable()
+export class DistributeLockService {
```
Comment (kkshyu) [inline src/utility/lock/distribute_lock.service.ts]: it should be `Distributed` not `Distribute`

File: src/runner/runner.module.ts [hunk 4]
```diff
@@ -1,23 +1,65 @@
...
+@Module({})
+export class RunnerModule {
+  static createRunnerProvider(workerName: RunnerType): Type<Runner> {
+    switch (workerName) {
+      case RunnerType.EXAMPLE_RUNNER:
+        return ExampleRunner;
+      default:
+        throw new Error(`Given runner name is not exists: ${workerName}`);
+    }
+  }
```
Comment (kkshyu) [inline src/runner/runner.module.ts]: it is annoying to add both RunnerType and here while adding a new runner. i think it is easy to forget to add both side. we can merge the switch case into enum. plz see the `runner.type`

File: src/runner/runner.type.ts [hunk 5]
```diff
@@ -0,0 +1,3 @@
+export enum RunnerType {
+  EXAMPLE_RUNNER = 'EXAMPLE_RUNNER',
```
Comment (kkshyu) [inline src/runner/runner.type.ts]: we can just specify the runner instance like this: 
```ts
export enum RunnerType {
    EXAMPLE_RUNNER = ExampleRunner,
}
```

File: src/tasker/tasker.module.ts [hunk 6]
```diff
@@ -1,10 +1,55 @@
...
+@Module({})
+export class TaskerModule {
+  static createTaskerProvider(workerName: TaskerType): Type<any> {
+    switch (workerName) {
+      case TaskerType.EXAMPLE_TASKER:
+        return ExampleTasker;
+      default:
+        throw new Error(`Given tasker name is not exists: ${workerName}`);
+    }
+  }
```
Comment (kkshyu) [inline src/tasker/tasker.module.ts]: the same as runner. it is better to merge into enum

File: src/tasker/tasker.module.ts [hunk 7]
```diff
@@ -1,10 +1,55 @@
...
+        ConfigModule.forRoot({
+          envFilePath: `${cwd()}/.env${nodeEnv ? `.${nodeEnv}` : ''}`,
+          isGlobal: true,
+        }),
+        BullModule.forRootAsync({
+          imports: [ConfigModule],
+          useFactory: (configService: ConfigService<{
+            QUEUE_REDIS_URI: string;
+          }>) => {
+            const queueRedisUri = configService.getOrThrow('QUEUE_REDIS_URI');
```
Comment (kkshyu) [inline src/tasker/tasker.module.ts:32]: the same as the above. it is better to use constant string variable

File: src/utility/lock/distribute_lock.service.ts [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+  constructor(
+    private cacheService: CacheService,
+  ) {}
+
+  async acquireLock(
+    key: string, acquirerUuid: string, expireTime?: number,
+  ): Promise<boolean> {
```
Comment (kkshyu) [inline src/utility/lock/distribute_lock.service.ts]: throw error if lock is already existed else return a uuid string

**Review by kkshyu — CHANGES_REQUESTED** [review]
there are some big discussions i would like to have:
1. putting `key: string` into LockModule constructor
2. returning string type from DistributedLockService.acquireLock(), throw error if there is an existed lock 
3. merge uuid&isComplete into lock in Runner -> acquire lock from DistrbutedLockService instead of generating uuid from runner
4. removing `key: string, acquirerUuid: string` from aquireLock, `key: string` from touchLock, `key: string` from releaseLock

there are some reasons:
1. it is much more readable and flexible for the following code

here are the following revisions:
1. make lockKey configurable
```ts
@Module({})
export class LockModule {
  static forRoot(lockKey: string): DynamicModule {
    return {
      module: LockModule,
      providers: [
        CacheService, 
        {
          provide: DistributedLockService,
          useValue: new DistributedLockService(lockKey),
        },
      ],
      exports: [DistributedLockService],
    };
  }
}
```

2. config the lockKey in RunnerModule
specify lockKey in RunnerModule is better than in Runner because we specify the runner type and name in RunnerModule, we should also provide a corresponded lock key for it.

```ts
class RunnerModule {
   ...
  static forRoot(options: {
    workerName: RunnerType; nodeEnv: string;
  }): DynamicModule {
    const { workerName, nodeEnv } = options;
    return {
      module: RunnerModule,
      imports: [
        ScheduleModule.forRoot(),
        TypeOrmModule.forRoot(AppDataSourceConfig),
        ...
        LockModule.forRoot("YOUR_LOCK_KEY")
      ]
    }   
}
```

3. Use the distributedLockService easily in Runner
```ts
export abstract class Runner {
  protected lock: string | null = null;
  protected readonly name: string;
  protected readonly interval: number;
  protected readonly lockService: DistributeLockService;
  constructor(
    name: string,
    interval: number,
    lockService: DistributeLockService
  ) {
    this.name = name;
    this.interval = interval;
    this.lockService = lockService;
  }

  abstract execute(): Promise<void>;

  async run(): Promise<void> {
    if (this.lock) {
      // check if lock is expired and extend
      try {
        // throw error if lock is expired or not match else extend the lock
        await this.lockService.touchLock(this.lock);
        return;
      } catch (error) {
        // there are some cases to here:
        // 1. lock is expired, but this runner is still running
        // 2. lock is expired and re-acquired somewhere, but this runner is still running
        // 3. lock is not expired, but runner is done without releasing lock
        // ....
        // all of them are weird cases -> notify us
        console.error(`weird cases for distributed lock: ${error}`);
        return;
      }
    }

    try {
      this.lock = await this.lockService.acquireLock(this.interval + 60000);
    } catch (error) {
      // cannot get lock from distributed lock service
      // 1. lock is acquired by other runner
      // 2. distributed lock service is down
      // ....
      // all of them are weird cases -> notify us
      console.error(`weird cases for acquiring distributed lock: ${error}`);
    }

    try {
      await this.execute();
    } catch (err) {
      console.error(err);
    } finally {
      await this.revoke();
    }
  }

  async revoke(): Promise<void> {
    // TODO: kill execute() if it is still running
    try {
      await this.lockService.releaseLock(this.lock);
    } catch (error) {
      // cannot release the lock
      console.error(`weird cases for releasing distributed lock: ${error}`);
    } finally {
      this.lock = null;
    }
  }
}

```

File: src/runner/runner.module.ts (hunk 1 above)
Comment (jackey8616) [reply src/runner/runner.module.ts:39]: image
  
K is `{ QUEUE_REDIS_URI: string; }`  
There will be no typo.

File: src/tasker/tasker.module.ts (hunk 7 above)
Comment (jackey8616) [reply src/tasker/tasker.module.ts:32]: As previous comment.

**Review by kkshyu — APPROVED** [review]

**Review by kkshyu — APPROVED** [review]

**Pull request merged**